graylog2-server/src/main/java/org/graylog2/indexer/retention/strategies/AbstractIndexRetentionStrategy.java

```diff
@@ -95,6 +90,22 @@ private void retainTimeBased(IndexSet indexSet, TimeBasedSizeOptimizingStrategyC
         }
     }
 
+    private DateTime getClosingDateBestEffort(String indexName) {
+        if (!indices.isClosed(indexName)) {
+            throw new IllegalStateException(f("Invalid request for closing date - index %s is not closed", indexName));
+        }
+        Optional<DateTime> closingDate = indices.indexClosingDate(indexName);
+        if (closingDate.isPresent()) {
+            return closingDate.get();
+        }
+        final IndexRangeStats indexRangeStats = indices.indexRangeStatsOfIndex(indexName);
```

Comment:
I'm not sure if we should pull in the timestamps from the log data itself.
So far this is strictly based on the time from the rotation and I think we shouldn't mix them.

For an alternative idea: If we don't have the `closingDate` can we use the `creationDate` and compare it against
the `indexLifeTimeHard()` instead?

What we could also do, is to make use of the fact that the creationDate of a newer index is kind of like the closingDate of the current one. e.g

```
index_4   creationDate  2022-12-05-...
index_3   creationDate  2022-12-04-...
index_2   creationDate  2022-12-03-...
```
the closingDate of `index_2` is most likely the creationDate of `index_3`

Reply:
Agreed - we shouldn't mingle message timestamps with index dates. Checking `creationDate` against `lifetimeHard` is a good idea.

Evaluating dates on other indices is a bit messy, due to having to parse and sort index names. I don't think it's worth adding that complexity, just for the transition period when there are still indices that have no `closingDate`. Instead I propose forcing retention as a fallback, when we cannot even find a `creationDate`.